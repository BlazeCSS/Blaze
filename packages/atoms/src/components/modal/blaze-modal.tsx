import { Component, Prop, Method, State, Element } from '@stencil/core';

@Component({
  tag: 'blaze-modal'
})
export class Modal {

  @Element() elem: HTMLElement;
  @Prop() ghost: boolean = false;
  @Prop() full: boolean = false;
  @Prop() open: boolean = false;
  @Prop() dismissible: boolean = false;
  @State() _isOpen: boolean = false;

  @Method()
  close() {
    this._isOpen = false;
  }

  @Method()
  show() {
    this._isOpen = true;
  }

  @Method()
  isOpen() {
    return this._isOpen;
  }

  componentWillLoad() {
    this._isOpen = this.open;
  }

  dismiss() {
    if (this.dismissible) this.close();
  }

  render() {
    const ghostClass = this.ghost ? `o-modal--ghost` : '';
    const fullClass = this.full ? `o-modal--full` : '';
    const modalIsOpenClass = this._isOpen ? 'o-modal--visible' : '';

    return (
      [
        <blaze-overlay aria-hidden open={this._isOpen} onClick={() => this.dismiss()}></blaze-overlay>,
        <div role="dialog" class={`o-modal ${ghostClass} ${fullClass} ${modalIsOpenClass}`}>
          {
            this.dismissible &&
            <button type="button" class="c-button c-button--close" onClick={() => this.close()}>
              &times;
            </button>
          }
          <slot />
        </div>
      ]
    );
  }
}
