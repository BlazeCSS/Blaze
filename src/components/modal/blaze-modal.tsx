import { h, Component, Event, EventEmitter, Prop, Method, Element } from '@stencil/core';

@Component({
  tag: 'blaze-modal',
})
export class Modal {
  @Element()
  elem: HTMLElement;

  @Prop()
  ghost: boolean = false;

  @Prop()
  full: boolean = false;

  @Prop({ mutable: true, reflect: true })
  open: boolean = false;

  @Prop()
  dismissible: boolean = false;

  @Prop()
  shouldClose: () => boolean;

  @Event({ eventName: 'close' })
  onClose: EventEmitter;

  @Method()
  async close() {
    if (!this.shouldClose || this.shouldClose()) {
      this.open = false;
      this.onClose.emit();
    }
  }

  @Method()
  async show() {
    this.open = true;
  }

  @Method()
  async isOpen() {
    return this.open;
  }

  dismiss() {
    if (this.dismissible) this.close();
  }

  render() {
    const ghostClass = this.ghost ? `o-modal--ghost` : '';
    const fullClass = this.full ? `o-modal--full` : '';
    const modalIsOpenClass = this.open ? 'o-modal--visible' : '';
    const overlayIsOpenClass = this.open ? 'c-overlay--visible' : '';

    return [
      <div aria-hidden onClick={() => this.dismiss()} class={`c-overlay c-overlay--fullpage ${overlayIsOpenClass}`} />,
      <div role="dialog" class={`o-modal ${ghostClass} ${fullClass} ${modalIsOpenClass}`}>
        {this.dismissible && (
          <button type="button" class="c-button c-button--close" onClick={() => this.close()}>
            &times;
          </button>
        )}
        <slot />
      </div>,
    ];
  }
}
