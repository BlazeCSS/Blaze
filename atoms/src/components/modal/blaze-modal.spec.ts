import { flush, render } from '@stencil/core/testing';
import { Modal } from './blaze-modal';

describe('Modal', () => {
  it('should build', () => {
    expect(new Modal()).toBeTruthy();
  });

  let element;

  const snapIt = (name, html) => {
    it(name, async () => {
      element = await render({
        components: [Modal],
        html
      });
      await flush(element);

      expect(element).toMatchSnapshot();
    });
  };

  snapIt(
    'renders correctly with no props',
    '<blaze-modal>default</blaze-modal>'
  );

  snapIt(
    'renders as open',
    '<blaze-modal open>default</blaze-modal>'
  );

  snapIt(
    'renders dismissible button and overlay',
    '<blaze-modal open dismissible>default</blaze-modal>'
  );

  it('show modal method', async () => {
    element = await render({
      components: [Modal],
      html: '<blaze-modal>default</blaze-modal>'
    });
    await flush(element);
    expect(element).toMatchSnapshot();

    element.show();
    await flush(element);
    expect(element.isOpen()).toEqual(true);
    expect(element).toMatchSnapshot();
  });

  it('close modal method', async () => {
    element = await render({
      components: [Modal],
      html: '<blaze-modal open>default</blaze-modal>'
    });
    await flush(element);
    expect(element).toMatchSnapshot();

    element.close();
    await flush(element);
    expect(element.isOpen()).toEqual(false);
    expect(element).toMatchSnapshot();
  });

  it('close button', async () => {
    element = await render({
      components: [Modal],
      html: '<blaze-modal open dismissible>default</blaze-modal>'
    });
    await flush(element);
    expect(element).toMatchSnapshot();

    element.querySelector('.c-button--close').click();
    await flush(element);
    expect(element.isOpen()).toEqual(false);
    expect(element).toMatchSnapshot();
  });

  it('close by clicking overlay', async () => {
    element = await render({
      components: [Modal],
      html: '<blaze-modal open dismissible>default</blaze-modal>'
    });
    await flush(element);
    expect(element).toMatchSnapshot();

    element.querySelector('.c-overlay--visible').click();
    await flush(element);
    expect(element.isOpen()).toEqual(false);
    expect(element).toMatchSnapshot();
  });
});
