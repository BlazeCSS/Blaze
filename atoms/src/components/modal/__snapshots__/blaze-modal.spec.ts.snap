// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal close button 1`] = `
<blaze-modal
  class="hydrated"
  dismissible=""
  open=""
>
  <div
    class="c-overlay c-overlay--visible c-overlay--dismissible"
  />
  <div
    class="o-modal"
  >
    <button
      class="c-button c-button--close"
      type="button"
    >
      ×
    </button>
    <!---->
    default
  </div>
</blaze-modal>
`;

exports[`Modal close button 2`] = `
<blaze-modal
  class="hydrated"
  dismissible=""
  open=""
>
  
  <div
    class="o-modal u-display-none"
  >
    <button
      class="c-button c-button--close"
      type="button"
    >
      ×
    </button>
    <!---->
    default
  </div>
</blaze-modal>
`;

exports[`Modal close by clicking overlay 1`] = `
<blaze-modal
  class="hydrated"
  dismissible=""
  open=""
>
  <div
    class="c-overlay c-overlay--visible c-overlay--dismissible"
  />
  <div
    class="o-modal"
  >
    <button
      class="c-button c-button--close"
      type="button"
    >
      ×
    </button>
    <!---->
    default
  </div>
</blaze-modal>
`;

exports[`Modal close by clicking overlay 2`] = `
<blaze-modal
  class="hydrated"
  dismissible=""
  open=""
>
  
  <div
    class="o-modal u-display-none"
  >
    <button
      class="c-button c-button--close"
      type="button"
    >
      ×
    </button>
    <!---->
    default
  </div>
</blaze-modal>
`;

exports[`Modal close modal method 1`] = `
<blaze-modal
  class="hydrated"
  open=""
>
  <div
    class="c-overlay c-overlay--visible"
  />
  <div
    class="o-modal"
  >
    
    <!---->
    default
  </div>
</blaze-modal>
`;

exports[`Modal close modal method 2`] = `
<blaze-modal
  class="hydrated"
  open=""
>
  
  <div
    class="o-modal u-display-none"
  >
    
    <!---->
    default
  </div>
</blaze-modal>
`;

exports[`Modal renders as open 1`] = `
<blaze-modal
  class="hydrated"
  open=""
>
  <div
    class="c-overlay c-overlay--visible"
  />
  <div
    class="o-modal"
  >
    
    <!---->
    default
  </div>
</blaze-modal>
`;

exports[`Modal renders correctly with no props 1`] = `
<blaze-modal
  class="hydrated"
>
  
  <div
    class="o-modal u-display-none"
  >
    
    <!---->
    default
  </div>
</blaze-modal>
`;

exports[`Modal renders dismissible button and overlay 1`] = `
<blaze-modal
  class="hydrated"
  dismissible=""
  open=""
>
  <div
    class="c-overlay c-overlay--visible c-overlay--dismissible"
  />
  <div
    class="o-modal"
  >
    <button
      class="c-button c-button--close"
      type="button"
    >
      ×
    </button>
    <!---->
    default
  </div>
</blaze-modal>
`;

exports[`Modal show modal method 1`] = `
<blaze-modal
  class="hydrated"
>
  
  <div
    class="o-modal u-display-none"
  >
    
    <!---->
    default
  </div>
</blaze-modal>
`;

exports[`Modal show modal method 2`] = `
<blaze-modal
  class="hydrated"
>
  <div
    class="c-overlay c-overlay--visible"
  />
  <div
    class="o-modal"
  >
    
    <!---->
    default
  </div>
</blaze-modal>
`;
